src/buf/slice.rs

```diff
@@ -127,36 +172,135 @@ impl<T: IoBuf> ops::Deref for Slice<T> {
     type Target = [u8];
 
     fn deref(&self) -> &[u8] {
-        &super::deref(&self.buf)[self.begin..self.end]
+        let buf_bytes = super::deref(&self.buf);
+        let end = cmp::min(self.end, buf_bytes.len());
+        &buf_bytes[self.begin..end]
     }
 }
 
 impl<T: IoBufMut> ops::DerefMut for Slice<T> {
     fn deref_mut(&mut self) -> &mut [u8] {
-        &mut super::deref_mut(&mut self.buf)[self.begin..self.end]
+        let buf_bytes = super::deref_mut(&mut self.buf);
+        let end = cmp::min(self.end, buf_bytes.len());
+        &mut buf_bytes[self.begin..end]
     }
 }
 
-unsafe impl<T: IoBuf> IoBuf for Slice<T> {
-    fn stable_ptr(&self) -> *const u8 {
-        ops::Deref::deref(self).as_ptr()
+impl<T: IoBuf> Slice<T> {
+    /// Returns a raw pointer to the vector’s buffer at the slice offset.
+    ///
+    /// This method is to be used by the `tokio-uring` runtime and it is not
+    /// expected for users to call it directly.
+    pub fn stable_ptr(&self) -> *const u8 {
```

Comment:
I'm not sure it's a good idea to combine inherent methods with the `Deref` impl, as per the API guidelines [C-DEREF], [C-SMART-PTR].

[C-SMART-PTR]: https://rust-lang.github.io/api-guidelines/predictability.html#c-smart-ptr
[C-DEREF]: https://rust-lang.github.io/api-guidelines/predictability.html#c-deref

As these methods are not meant to be called by the API user, perhaps they should be turned into associated functions and/or made private.

On the other hand, the `Deref`/`DerefMut` impls for `Slice` could be considered for removal in favor of explicit accessor methods like `.as_bytes()`/`.as_mut_bytes()`. Especially if we'd want to equip `Slice` with more non-slicey methods for appending to buffers, like `.put_bytes`, `.put_u32_le`, etc.

Reply:
> explicit accessor methods like `.as_bytes()` / `.as_mut_bytes()`

... or a `std::ops::Index` impl, as this is what the users would need the most from a buffer, and `[..]` is a concise adapter syntax for anything else where buffer contents would be used as a byte slice.

Reply:
I have realized that the `IoBuf`/`IoBufMut`-cum-inherent-`Slice` methods do not need to be public, so I've made them `pub(crate)`. But the above comments still apply to the other inherent methods of `Slice`.